sylvia/src/multitest.rs

```diff
@@ -1,62 +1,100 @@
 use std::cell::{Ref, RefCell, RefMut};
+use std::fmt::{Debug, Display};
 use std::marker::PhantomData;
+use std::ops::Deref;
 
-use cosmwasm_std::{Addr, BlockInfo, Coin};
-use cw_multi_test::Executor;
+use cosmwasm_std::{Addr, Api, Coin, CustomQuery, Empty, GovMsg, IbcMsg, IbcQuery, Storage};
+use cw_multi_test::{
+    BankKeeper, DistributionKeeper, Executor, FailingModule, Router, StakeKeeper, WasmKeeper,
+};
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
 use serde::Serialize;
 
-#[derive(Default)]
-pub struct App {
-    pub app: RefCell<cw_multi_test::App>,
+pub struct App<MtApp> {
+    app: RefCell<MtApp>,
 }
 
-impl App {
-    pub fn new(app: cw_multi_test::App) -> Self {
-        Self {
-            app: RefCell::new(app),
-        }
+impl<MtApp> Default for App<MtApp>
+where
+    MtApp: Default,
+{
+    fn default() -> Self {
+        Self::new(MtApp::default())
     }
+}
 
-    pub fn app(&self) -> Ref<'_, cw_multi_test::App> {
-        Ref::map(self.app.borrow(), |app| app)
+impl<MtApp> Deref for App<MtApp> {
+    type Target = RefCell<MtApp>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.app
     }
+}
 
-    pub fn app_mut(&self) -> RefMut<'_, cw_multi_test::App> {
-        RefMut::map(self.app.borrow_mut(), |app| app)
+impl<ExecC, QueryC> App<cw_multi_test::BasicApp<ExecC, QueryC>> {
+    /// Creates new default `App` implementation working with customized exec and query messages.
+    pub fn custom<F>(init_fn: F) -> Self
+    where
+        ExecC: Debug + Clone + PartialEq + JsonSchema + DeserializeOwned + 'static,
+        QueryC: Debug + CustomQuery + DeserializeOwned + 'static,
+        F: FnOnce(
+            &mut Router<
+                BankKeeper,
+                FailingModule<ExecC, QueryC, Empty>,
+                WasmKeeper<ExecC, QueryC>,
+                StakeKeeper,
+                DistributionKeeper,
+                FailingModule<IbcMsg, IbcQuery, Empty>,
+                FailingModule<GovMsg, Empty, Empty>,
+            >,
+            &dyn Api,
+            &mut dyn Storage,
+        ),
+    {
+        App {
+            app: RefCell::new(cw_multi_test::custom_app(init_fn)),
+        }
     }
+}
 
-    pub fn block_info(&self) -> BlockInfo {
-        self.app.borrow().block_info()
+impl<MtApp> App<MtApp> {
+    pub fn new(app: MtApp) -> Self {
+        Self {
+            app: RefCell::new(app),
+        }
     }
 
-    pub fn set_block(&self, block: BlockInfo) {
-        self.app.borrow_mut().set_block(block)
+    pub fn app(&self) -> Ref<'_, MtApp> {
+        Ref::map(self.app.borrow(), |app| app)
     }
 
-    pub fn update_block<F: Fn(&mut BlockInfo)>(&self, action: F) {
-        self.app.borrow_mut().update_block(action)
+    pub fn app_mut(&self) -> RefMut<'_, MtApp> {
+        RefMut::map(self.app.borrow_mut(), |app| app)
     }
 }
 
 #[must_use]
-pub struct ExecProxy<'a, 'app, Error, Msg>
+pub struct ExecProxy<'a, 'app, Error, Msg, MtApp, ExecC = Empty>
```

Comment:
I am struggling if it is a good idea to have this `Empty` default for `ExecC` - does it save much? If we have it - why not make `MtApp` default to `mt::BasicApp<Empty>`? I kind of see that `MtApp` have this additional `QueryC` you will sometimes change... Therefore I am not super sure. I think I would just get rid of default here as it saves you a bit right now, but in the future, you will anyway always spell this whole type in generated code, and the type is not to be spelled out by the user anyway. Removing the default here would save you some trouble working with errors of "`MtApp doesn't implement the `Executor<ExecC>`, but it implements `Executor<SomeOther`", bla bla bla, which would happen if you would have non-default `MtApp` not matching `Empty` `ExecC`, and you forget to specify `ExecC`. If you undefault it, you will have an error of missing generic and things should be a bit easier to maintain.